### PR TITLE
Add support for custom URLs for email verification

### LINKED ISSUE #22561: Allow a Custom URL in the new POST /register endpoint
### Describe the Improvement

The invite endpoint allows for setting an `invite_url` property (with a `USER_INVITE_URL_ALLOW_LIST` variable) for handling invite logic inside of an external application. 

The password reset endpoint is similar with a `reset_url` property and `PASSWORD_RESET_URL_ALLOW_LIST` variable. 

The new public registration endpiont does not have an equivalent. I propose `verify_url` or `verification_url` with a variable of `PUBLIC_REGISTRATION_URL_ALLOW_LIST`.

### PR DESCRIPTION
**Pull request by DanielBiegler**
## Scope

What's changed:

- New Env Var `USER_REGISTER_URL_ALLOW_LIST` similar to `USER_INVITE_URL_ALLOW_LIST`
- Initial doc (misses link to API-doc since it doesnt exist yet)
- Register payload accepts a new Url parameter
- If said parameter is set, you must configure the env var (just like with invites)
- If said parameter is set and valid, the verification URL inside of the email is now your specified url allowing you to build a custom verification page outside the data studio

## Potential Risks / Drawbacks

- This follows the logic how invites work but we could change it to do the validation inside of the joi schema

## Review Notes / Questions

- We might want to change the env var to `USER_VERIFY_URL_ALLOW_LIST` to make it more clear that this is pertaining to verification and not registration for example but thats a minor thing

---

Fixes #22561

### CODE REVIEW

File: docs/self-hosted/config-options.md [hunk 1]
```diff
@@ -357,6 +357,7 @@ Redis is required when you run Directus load balanced across multiple containers
 | `REGISTER_STALL_TIME`               | The duration in milliseconds that a registration request will be stalled for, and it should be greater than the time taken for a registration request with an already registered email               | `750`                     |
 | `PASSWORD_RESET_URL_ALLOW_LIST`     | List of URLs that can be used [as `reset_url` in /password/request](/reference/authentication#request-password-reset)                                                                                | --                        |
 | `USER_INVITE_URL_ALLOW_LIST`        | List of URLs that can be used [as `invite_url` in /users/invite](/reference/system/users#invite-a-new-user)                                                                                          | --                        |
+| `USER_REGISTER_URL_ALLOW_LIST`      | List of URLs that can be used as `verification_url` in /users/register                                                                                                                               | --                        |
```
Comment (DanielBiegler) [inline docs/self-hosted/config-options.md:360]: @phazonoverload we might wanna add the docs in this PR, then we can link to the correct section in one swoop. I'll leave this as draft for now

**Comment by phazonoverload** [comment]
Feel free to merge - the rest isn't documented anyway. I'm working on it and will incorporate this ;)

**Review by paescuj — APPROVED** [review]
🎉

**Comment by shubham-am-i** [comment]
Hi Team,

I'm encountering an issue with the recent update that adds support for custom URLs in email verification. Despite the fix being merged and the verification_url now included, I still receive an error indicating an invalid payload. Here are the details:

`{
    "errors": [
        {
            "message": "Invalid payload. \"verification_url\" is not allowed.",
            "extensions": {
                "code": "INVALID_PAYLOAD",
                "reason": "\"verification_url\" is not allowed"
            }
        }
    ]
}`

Context
This issue arises when attempting to utilize the new verification_url parameter in the registration payload. According to the merged PR (#22565), the payload should accept a new Url parameter, provided that the appropriate environment variable (USER_REGISTER_URL_ALLOW_LIST) is configured.

Steps Taken
1. Environment Variable Configuration:
Set the USER_REGISTER_URL_ALLOW_LIST environment variable, mirroring the logic used for USER_INVITE_URL_ALLOW_LIST.
2. Verified that the environment variable is correctly loaded by the application.
3. Payload Structure:
Constructed the registration payload with the verification_url parameter, ensuring it adheres to the expected format.
Verification:
4. Checked the codebase to ensure the recent changes from PR #22565 are present and correctly integrated.


Potential Issues
1. Schema Validation: There might be an oversight in the Joi schema validation, where the new verification_url parameter isn't yet recognized or allowed.
2. Environment Variable Misconfiguration: Although unlikely, there could be a misconfiguration or misinterpretation of the environment variable setup.
3. Documentation Gap: The initial documentation mentions a missing link to the API documentation, which might be crucial for understanding the exact requirements and constraints of the new parameter.


Request for Assistance
Could you please review the schema validation logic to confirm that the verification_url parameter is correctly permitted? Additionally, any updates or clarifications in the documentation regarding the usage and setup of this new feature would be greatly appreciated.

Thank you for your attention to this matter. Looking forward to your guidance on resolving this issue.

**Comment by DanielBiegler** [comment]
Hey @shubham-am-i it sounds like youre not on the latest version `v10.11.2` - it is confirmed to work.

For example: 

```
USER_REGISTER_URL_ALLOW_LIST=http://asd.de
```

And then an API request like the following:

```bash
curljson \
-d '{ "email": "test@example.com", "password": "test", "verification_url": "http://asd.de" }' \
<YOUR-DIRECTUS-URL-HERE>/users/register
```

works and correctly uses the verification URL inside of the email.

Please verify that youre on the correct Directus version and are POSTing to the correct URL `/users/register`

If you have further troubles please provide your settings and api request in a new issue instead of this PR - Thank you! :)

**Pull request merged**